File: BinarySubarraysWithSum.java
import java.util.HashMap;
import java.util.Map;

public class BinarySubarraysWithSum {
    public int numSubarraysWithSum(int[] nums, int goal) {
        int result = 0;
        int sum = 0;
        Map<Integer, Integer> order = new HashMap<>();
        order.put(0, 1);
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            result += order.getOrDefault(sum - goal, 0);
            order.put(sum, order.getOrDefault(sum, 0) + 1);
        }
        return result;
    }

    public static void main(String[] args) {
        BinarySubarraysWithSum solution = new BinarySubarraysWithSum();
        int[] nums = {1, 0, 1, 0, 1}; // Example input array
        int goal = 2; // Example goal
        int result = solution.numSubarraysWithSum(nums, goal);
        System.out.println("Number of subarrays with sum equal to goal: " + result);
    }
}

File: BinarySubarraysWithSum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        int result = 0;
        int sum = 0;
        unordered_map<int, int> order;
        order[0] = 1;

        for (int i = 0; i < nums.size(); ++i) {
            sum += nums[i];
            result += order[sum - goal];
            order[sum]++;
        }

        return result;
    }
};

int main() {
    Solution solution;
 vector<int> nums = {1, 0, 1, 0, 1}; // Example input
    int goal = 2; // Example goal
    cout << "Number of subarrays: " << solution.numSubarraysWithSum(nums, goal) << endl;
    return 0;
}